x-pack/plugins/apm/server/lib/services/get_services/get_services_items.ts

```diff
@@ -19,72 +20,315 @@ import {
+
+type ServicesItemsProjection = ReturnType<typeof getServicesProjection>;
+
+interface AggregationParams {
+  setup: ServicesItemsSetup;
+  projection: ServicesItemsProjection;
+}
 
-  const params = mergeProjection(projection, {
-    body: {
+const getTransactionDurationAvg = async ({
+  setup,
+  projection,
+}: AggregationParams) => {
+  const { client } = setup;
+
+  const response = await client.search(
+    mergeProjection(projection, {
       size: 0,
-      aggs: {
-        services: {
-          terms: {
-            ...projection.body.aggs.services.terms,
-            size: 500,
+      body: {
+        query: {
+          bool: {
+            filter: projection.body.query.bool.filter.concat({
+              term: {
+                [PROCESSOR_EVENT]: 'transaction',
+              },
+            }),
           },
-          aggs: {
-            avg: {
-              avg: { field: TRANSACTION_DURATION },
+        },
+        aggs: {
+          services: {
+            terms: {
+              ...projection.body.aggs.services.terms,
+              size: MAX_NUMBER_OF_SERVICES,
             },
-            agents: {
-              terms: { field: AGENT_NAME, size: 1 },
+            aggs: {
+              average: {
+                avg: {
+                  field: TRANSACTION_DURATION,
+                },
+              },
+            },
+          },
+        },
+      },
+    })
+  );
+
+  const { aggregations } = response;
+
+  if (!aggregations) {
+    return [];
+  }
+
+  return aggregations.services.buckets.map((bucket) => ({
+    name: bucket.key as string,
+    value: bucket.average.value,
+  }));
+};
+
+const getAgentName = async ({ setup, projection }: AggregationParams) => {
+  const response = await setup.client.search(
+    mergeProjection(projection, {
+      body: {
+        query: {
+          bool: {
+            filter: [
+              ...projection.body.query.bool.filter,
+              {
+                terms: {
+                  [PROCESSOR_EVENT]: ['metric', 'error', 'transaction'],
+                },
+              },
+            ],
+          },
+        },
+        aggs: {
+          services: {
+            terms: {
+              ...projection.body.aggs.services.terms,
+              size: MAX_NUMBER_OF_SERVICES,
             },
-            events: {
-              terms: { field: PROCESSOR_EVENT },
+            aggs: {
+              agent_name: {
+                top_hits: {
+                  _source: [AGENT_NAME],
+                  size: 1,
+                },
+              },
             },
-            environments: {
-              terms: { field: SERVICE_ENVIRONMENT },
+          },
+        },
+        size: 0,
+      },
+    })
+  );
+
+  const { aggregations } = response;
+
+  if (!aggregations) {
+    return [];
+  }
+
+  return aggregations.services.buckets.map((bucket) => ({
+    name: bucket.key as string,
+    value: (bucket.agent_name.hits.hits[0]?._source as {
```

Comment:
I'm not sure if it's possible, but can't we define the response type we expect when calling `setup.client.search<T>`? Then we could avoid this.

Reply:
We can, but then we have to create a separate variable for params and use `client.search<{ agent: { name: string } }, typeof params>`. Which means we lose type checks and autocompletion when we create the request.

Reply:
(As soon as you start defining generic arguments TS will no longer infer others, AFAIK).